src/lib/agent/PolykeyAgent.ts

```diff
@@ -402,33 +471,52 @@ class PolykeyAgent {
   private async listVaults(nodePath: string) {
     const pk = this.getPolyKey(nodePath);
     const vaultNames = pk.vaultManager.listVaults();
-    return ListVaultsResponseMessage.encode({ vaultNames }).finish();
+    return ListVaultsResponseMessage.encodeDelimited({ vaultNames }).finish();
+  }
+  private async scanVaultNames(nodePath: string, request: Uint8Array) {
+    const { publicKey } = ScanVaultNamesRequestMessage.decodeDelimited(request);
+    const pk = this.getPolyKey(nodePath);
+    const vaultNames = await pk.vaultManager.scanVaultNames(publicKey);
+    return ScanVaultNamesResponseMessage.encodeDelimited({ vaultNames }).finish();
   }
   private async newVault(nodePath: string, request: Uint8Array) {
-    const { vaultName } = NewVaultRequestMessage.decode(request);
+    const { vaultName } = NewVaultRequestMessage.decodeDelimited(request);
     const pk = this.getPolyKey(nodePath);
     await pk.vaultManager.createVault(vaultName);
-    return NewVaultResponseMessage.encode({ successful: true }).finish();
+    return NewVaultResponseMessage.encodeDelimited({ successful: true }).finish();
+  }
+  private async cloneVault(nodePath: string, request: Uint8Array) {
+    const { vaultName, publicKey } = CloneVaultRequestMessage.decodeDelimited(request);
+    const pk = this.getPolyKey(nodePath);
+    await pk.vaultManager.cloneVault(vaultName, publicKey);
+    return CloneVaultResponseMessage.encodeDelimited({ successful: true }).finish();
+  }
+  private async pullVault(nodePath: string, request: Uint8Array) {
+    const { vaultName, publicKey } = PullVaultRequestMessage.decodeDelimited(request);
+    const pk = this.getPolyKey(nodePath);
+    const vault = pk.vaultManager.getVault(vaultName);
+    vault.pullVault(publicKey);
+    return PullVaultResponseMessage.encodeDelimited({ successful: true }).finish();
   }
   private async destroyVault(nodePath: string, request: Uint8Array) {
-    const { vaultName } = DestroyVaultRequestMessage.decode(request);
+    const { vaultName } = DestroyVaultRequestMessage.decodeDelimited(request);
```

Comment:
Can you explain the difference between `decode` and `decodeDelimited` and why?

Reply:
decodeDemilited just validates how large the message was at encode message and throws if its not exactly that long. It helps when aggregating messages from TCP sockets as packets can be split up sometimes.